pickle_server/exceptions.py

```diff
@@ -0,0 +1,72 @@
+import re
+from ninja.errors import AuthenticationError
+from django.core.exceptions import PermissionDenied, ObjectDoesNotExist
+from django.core.exceptions import ValidationError as DjangoValidationError
+from ninja.errors import ValidationError as NinjaValidationError
+
+from typing import Optional
+
+from ninja import Schema
+from ninja.responses import codes_2xx, codes_4xx
+
+
+class APIError(Exception):
+    status_code = 500
+    detail = "Internal Server Error"
+
+    def __init__(self, detail=None, status_code=None):
+        if detail:
+            self.detail = detail
+        if status_code is not None:
+            self.status_code = status_code
+
+    def __str__(self):
+        return self.detail
+
+
+class ResponseDTO(Schema):
+    success: bool
+    error_type: Optional[str] = None
+    detail: Optional[str] = ""
+
+    @classmethod
+    def from_status_code(cls, status_code: int, detail: Optional[str] = ""):
+        if status_code in codes_2xx:
+            return cls(success=True, detail=detail)
+        elif status_code in codes_4xx:
+            return cls(success=False, detail=detail)
+        else:
+            raise NotImplementedError
+
+
+def camel_to_snake(name):
+    p = re.compile(r"(?<!^)(?=[A-Z])")
+    return p.sub("_", name).lower()
+
+
+def api_exception_response(request, exc):
+    status_code = 500
+    error_msg = str(exc)
+    error_type = camel_to_snake(exc.__class__.__name__)
+    if isinstance(exc, APIError):
+        status_code = exc.status_code
+    elif isinstance(exc, DjangoValidationError):
+        status_code = 400
+        error_msg = exc.message
+    elif isinstance(exc, NinjaValidationError):
+        status_code = 422
+        error_msg = exc.errors[0]["msg"]
+    elif isinstance(exc, PermissionDenied):
+        status_code = 403
+    elif isinstance(exc, ObjectDoesNotExist):
+        status_code = 404
+    elif isinstance(exc, AuthenticationError):
+        status_code = 401
```

Comment:
p5) 취향의 영약이긴 하나 이런 경우에는 패턴 매칭도 이쁘더라구요 (근데 우리 3.11 맞나요? python-version 파일에 pickle-server 라고 되어있던데 처음보는 양식이라 ㅠ 잘 몰라서 그러는데 보통 python-version 파일에 3.11 이런 식으로 버전 써놓지 않아요??)
```suggestion
    match exc:
        case APIError():
            status_code = exc.status_code
        case DjangoValidationError():
            status_code = 400
            error_msg = exc.message
        ...
```

Reply:
.python-version파일은 제가 파이썬 버전 관리에 쓰는 [pyenv](https://github.com/pyenv/pyenv)에서 사용하는 양식인데, virtual env이름이 들어가서 `pickle-server`와 같이 되어있네요. 로컬에서만 사용되는 virtualenv name이어서 이렇게만 두면 안됐는데 저의 실수입니다..
1. `.python-version`을 `gitignore`에 추가해두거나
2. `pickle-server` virtualenv 셋업하는 법을 README에 적어놓을게요

match case는 아시다싶이 3.12부터 지원하는데, 아직 지원안하는 종속성  (라이브러리 기타 등등)이 있을까봐 보수적으로 잡아서 3.11로 세팅을 해두었습니다